sources/hachiko/src/main.rs

```diff
@@ -0,0 +1,71 @@
+mod config;
+mod error;
+use crate::config::Config;
+use crate::error::Result;
+use imdsclient::ImdsClient;
+use log::LevelFilter;
+use log::{error, info, trace, warn};
+use simplelog::{Config as LogConfig, SimpleLogger};
+use snafu::{OptionExt, ResultExt};
+use std::fs;
+use tokio::time::{sleep, Duration};
+
+// Path to config file containing host's autoscaling.should-wait setting
+const CONFIG_PATH: &str = "/etc/hachiko.toml";
+
+/// Uses an imdsclient function to fetch the lifecyclestate of the host
+async fn get_lifecycle_state(client: &mut ImdsClient) -> Result<String> {
+    client
+        .fetch_lifecycle_state()
+        .await
+        .context(error::ImdsRequestSnafu)?
+        .context(error::ImdsNoneSnafu {
+            what: "instance-id",
+        })
+}
+
+/// Continuously fetches lifecycle state of the host.
+/// Returns from function when host is "InService"
+async fn wait_until_inservice() -> Result<()> {
+    let mut client = ImdsClient::new();
+    let mut lifecycle_state = get_lifecycle_state(&mut client).await?;
+    info!("Lifecycle state is {}", lifecycle_state);
+    while lifecycle_state.ne("InService") {
+        lifecycle_state = get_lifecycle_state(&mut client).await?;
+        info!("Lifecycle state is {} ... waiting.", lifecycle_state);
+        sleep(Duration::from_secs(5)).await;
+    }
+    info!("Lifecycle state is {} ... exiting.", lifecycle_state);
+    Ok(())
+}
+
+/// Parses config file for autoscaling.should-wait setting.
+async fn get_config() -> Result<Config> {
+    Config::from_file(CONFIG_PATH)
+}
+
+/// If autoscaling.should-wait is true, calls waiting function.
+/// Writes marker file to prevent service from unexpectedly re-running.
+#[tokio::main]
+async fn main() {
+    SimpleLogger::init(LevelFilter::Info, LogConfig::default())
+    .expect("unable to configure logger");
```

Comment:
Let's use the `.context(snafu_error_here)` to provide context rather than using `expect()`